Game/Source/GDKTestGyms/BenchmarkGymGameModeBase.cpp

```diff
@@ -1054,6 +1063,83 @@ void ABenchmarkGymGameModeBase::CheckVelocityForPlayerMovement()
 	}
 }
 
+void ABenchmarkGymGameModeBase::OutputPlayerDensity()
+{
+	// Outputs the count that each NPC and Simulated Player falls into each of the QBI-F bucket types. This is not performant but
+	// is only used for debugging purposes currently and isn't enabled by default.
+	FTimerHandle CountTimer;
+	GetWorld()->GetTimerManager().SetTimer(
+		CountTimer,
+		[WeakThis = TWeakObjectPtr<ABenchmarkGymGameModeBase>(this)]() {
+		if (ABenchmarkGymGameModeBase* GameMode = WeakThis.Get())
+		{
+			USpatialNetDriver* SpatialDriver = Cast<USpatialNetDriver>(GameMode->GetNetDriver());
+
+			TArray<AActor*> PlayerControllers, PlayerCharacters, NPCs, AllCharacters;
+			UGameplayStatics::GetAllActorsOfClass(GameMode->GetWorld(), GameMode->SimulatedPlayerControllerClass, PlayerControllers);
+			UGameplayStatics::GetAllActorsOfClass(GameMode->GetWorld(), GameMode->SimulatedPawnClass, PlayerCharacters);
+			UGameplayStatics::GetAllActorsOfClass(GameMode->GetWorld(), GameMode->NPCClass, NPCs);
+			AllCharacters = PlayerCharacters;
+			AllCharacters.Append(NPCs);
+
+			const USpatialGDKSettings* SpatialGDKSettings = GetDefault<USpatialGDKSettings>();
+			TArray<float> NCDDistanceRatios;
+			NCDDistanceRatios.Push(SpatialGDKSettings->FullFrequencyNetCullDistanceRatio);
+			for (auto i : SpatialGDKSettings->InterestRangeFrequencyPairs)
+			{
+				NCDDistanceRatios.Push(i.DistanceRatio);
+			}
+			NCDDistanceRatios.Push(MAX_FLT);
+			NCDDistanceRatios.Sort();
+
+			TArray<int> TotalCountPerBucket;
+			TArray<int> CountPerBucket;
+			TotalCountPerBucket.Init(0, NCDDistanceRatios.Num());
+
+			for (AActor* PlayerController : PlayerControllers)
+			{
+				CountPerBucket.Init(0, NCDDistanceRatios.Num());
+
+				FVector Pos = SpatialDriver->GetActorChannelByEntityId(SpatialDriver->GetActorEntityId(*PlayerController))->GetLastUpdatedSpatialPosition();
+				for (AActor* Character : AllCharacters)
+				{
+					FVector OtherPos = SpatialDriver->GetActorChannelByEntityId(SpatialDriver->GetActorEntityId(*Character))->GetLastUpdatedSpatialPosition();
+					float Dist = FVector::Distance(Pos, OtherPos);
+					float NCD = FMath::Sqrt(Character->NetCullDistanceSquared);
+					for (int i = 0; i < NCDDistanceRatios.Num(); ++i)
+					{
+						if (Dist < NCDDistanceRatios[i] * NCD)
+						{
+							CountPerBucket[i]++;
+							TotalCountPerBucket[i]++;
+							break;
+						}
+					}
+				}
+
+				int TotalCount = 0;
+				FString CountsAsString;
+				for (int Count : CountPerBucket)
+				{
+					CountsAsString += FString::Format(TEXT(" {0}"), { Count });
+					TotalCount += Count;
+				}
+
+				UE_LOG(LogBenchmarkGymGameModeBase, Log, TEXT("Density: %s (%d)"), *CountsAsString, TotalCount);
+			}
+
+			int TotalCount = 0;
+			FString CountsAsString;
+			for (int Count : TotalCountPerBucket)
```

Comment:
Possibly could use a lambda for printing out this data as we kinda use do the same logic twice except for slight string change... but just if you like...